feat: Implement internal audio/video recording for devices >= Android Q

sirmordred opened the pull request:
…Android.Q

REQUIRED STEPS to activate screenrecord feature:
- adb shell pm grant io.appium.settings android.permission.RECORD_AUDIO
- adb shell appops set io.appium.settings PROJECT_MEDIA allow

TO START SCREENRECORD:
- adb shell am start -n "io.appium.settings/io.appium.settings.Settings" -a io.appium.settings.recording.ACTION_START --es filename abc.mp4

TO STOP SCREENRECORD:
- adb shell am start -n "io.appium.settings/io.appium.settings.Settings" -a io.appium.settings.recording.ACTION_STOP

TO PULL RECORDED MP4 FILE:
- adb pull /storage/emulated/0/Android/data/io.appium.settings/files/abc.mp4 abc.mp4

sirmordred commented:
cc @KazuCocoa @mykola-mokhnach

sirmordred commented:
addressing https://github.com/appium/appium/issues/13177

sirmordred commented:
It works flawlessly as far as I tested it on my android device with latest appium, note: io.appium.settings apk which includes this commit must be built and installed and above 3 adb commands need to be issued first, here is sample test script including recording https://gist.github.com/sirmordred/4c505ddce96f7513bfbe49d38ef041be

KazuCocoa reviewed (commented):
Thank you.
Is there any reason why you did not add this feature in https://github.com/appium/appium-uiautomator2-server , for example, as a `mobile:` command? https://github.com/appium/appium-uiautomator2-driver
(kind of a basic question as my out of curiosity)

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+        return null;
+    }
+
+    @SuppressLint("NewApi")
```
I'd rather add an explicit check for the api version

sirmordred replied:
same https://github.com/appium/io.appium.settings/pull/88#discussion_r878926712

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+    /**
+     * start screen recording
+     */
+    @SuppressLint("NewApi")
```
same as above

sirmordred replied:
same https://github.com/appium/io.appium.settings/pull/88#discussion_r878926712

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+                int height = metrics.heightPixels;
+                if (width > height) {
+                    // Horizontal
+                    final float scale_x = width / 1920f;
```
can scale be made configurable?

mykola-mokhnach replied:
please use camelCase for variable names

sirmordred replied:
fixed camelCase and yeah scaling parameters can be passed through adb command initiated intent and with that way these settings can be configurable, the scaling adjustments are just for portrait and landscape phone modes

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+                    // Horizontal
+                    final float scale_x = width / 1920f;
+                    final float scale_y = height / 1080f;
+                    final float scale = Math.max(scale_x,  scale_y);
```
I can observe code duplication in both `if` branches

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+                Log.v(TAG, String.format("startRecording:(%d,%d)(%d,%d)",
+                        metrics.widthPixels, metrics.heightPixels, width, height));
+                String outputFilePath = intent.getStringExtra(ACTION_RECORDING_FILENAME);
+                if (outputFilePath != null) {
```
shall we check if recording is already running?
shall we log an error if no output file path is provided?

sirmordred replied:
we are checking status of recording with nullability currently but i can change to helper method(isRunning) 👍  and `outputFilePath` null check in here is just an extra check to solidify code, it must not be null in all conditions because we are supplying default file name in here https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/Settings.java#L107 but again i can add extra check 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    protected int calcBitRate(final int frameRate, int width, int height) {
+        final int bitrate = (int)(0.25f * frameRate * width * height);
```
what is the meaning of these magic numbers? could they be made configurable?

sirmordred replied:
done, move to constants, 0.25f is just BPP for calculating bitrate, refer to https://www.intopix.com/blogs/post/How-to-define-the-compression-rate-according-to-bpp-or-bps#how-to-calculate i also took it from somewhere stackoverflow or google github video encoding samples

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    private void stopScreenRecord() {
+        if (recorderThread != null) {
+            recorderThread.stopRecording();
```
should we assign recorderThread to null?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.videoBitrate = videoBitrate;
+        this.sampleRate = 44100;
```
Please move magic numbers to constants

sirmordred replied:
done, moved to constants

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+    }
+
+    public void stopRecording() {
+        stopped = true;
```
should recordingThread be stopped/nullified here?

sirmordred replied:
we are just calling this function on RecorderService stopScreenRecord() method and with the latest update, we now nullifying recorderThread object

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+                MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
+        encoderFormat.setInteger(MediaFormat.KEY_BIT_RATE, videoBitrate);
+        encoderFormat.setInteger(MediaFormat.KEY_FRAME_RATE, frameRate);
+        encoderFormat.setInteger(MediaFormat.KEY_REPEAT_PREVIOUS_FRAME_AFTER, 1000000);
```
Please move magic numbers to constants

sirmordred replied:
done, moved to constants

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+        // Encoded video resolution matches virtual display. TODO set channelCount 2 try stereo quality
+        MediaFormat encoderFormat = MediaFormat.createAudioFormat(MediaFormat.MIMETYPE_AUDIO_AAC,
+                sampleRate, 1);
+        encoderFormat.setInteger(MediaFormat.KEY_BIT_RATE, 64000);
```
magic numbers

sirmordred replied:
done, moved to constants

KazuCocoa commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
+        // adb shell appops set io.appium.settings PROJECT_MEDIA allow
+
+        // TO START SCREENRECORD:
+        // adb shell am start -n "io.appium.settings/io.appium.settings.Settings" -a io.appium.settings.recording.ACTION_START --es "recording_filename" "abc.mp4"
```
Should this be start activity? I wonder if this feature can be via `broadcast`. (I haven't dug in the recording feature, so I might miss something)

sirmordred replied:
i tried the initiating recording with `broadcast` but since i had to trigger startActivityForResult() method supplied with mediaprojection intent object (see https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/Settings.java#L108) i had to use Settings activity

sirmordred replied:
but i will try to dig more on this

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+                    audioEncoder.releaseOutputBuffer(encoderStatus, false);
+
+                    if ((bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
+                        Log.e(TAG, "buffer eos");
```
is this actually an error?

sirmordred replied:
done, converted to verbose

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+                if (videoTrackIndex >= 0 && audioTrackIndex < 0)
+                    continue; // wait for audio config before processing any video data frames
+
+                encoderStatus = videoEncoder.dequeueOutputBuffer(bufferInfo, 10000);
```
what is this magic number?

sirmordred replied:
done, moved to constants, it is just for timeout value while dequeuing buffer, it blocks buffering at most "timeoutUs" microseconds so no hang

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+                    }
+                }
+
+                if (videoTrackIndex >= 0 && audioTrackIndex < 0)
```
Please always encode condition body into curly brackets

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+                    audioEncoder = null;
+                }
+            } catch (Exception e) {
+                Log.w(TAG, "Error releasing resources", e);
```
this might not work as expected if an exception happens, because then all further resources won't be released properly

sirmordred replied:
> Thank you. Is there any reason why you did not add this feature in https://github.com/appium/appium-uiautomator2-server , for example, as a `mobile:` command? https://github.com/appium/appium-uiautomator2-driver (kind of a basic question as my out of curiosity)
> 
> [updated]
> 
> ah, this was for the below comment in the linked issue?
> 
> > Which means one can simply install any sound recording app, control it using UIA2 to start/stop recording process and be happy about it %)

yeah since recording on android via public android APIs are only available on application level, thats why i targeted io.appium.settings app and for future direction, i think yeah the start/stop recording commands can be folded into much less lenghty commands like `mobile: startRecordingEtc.`

sirmordred replied:
done, my bad sorry, separated try-catches

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
             LocationTracker.getInstance().start(this);
         }
 
+        // REQUIRED STEPS to activate screenrecord feature:
```
Please add this information into README

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
+
+        // handle screen recording
+        Intent openingIntent = getIntent();
+        if (openingIntent != null && openingIntent.getAction() != null) {
```
I'd rather move this block into a separate helper method

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
+
+                    if (recordingFilename == null || recordingFilename.isEmpty()
+                            || !recordingFilename.endsWith(".mp4")) {
+                        recordingFilename = "AppiumScreenRecord.mp4";
```
Please move magic strings into constants

sirmordred replied:
done, moved to constants

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
     }
 
-    private void registerSettingsReceivers(List<Class<? extends BroadcastReceiver>> receiverClasses) {
+    private boolean checkRecordingPermission() {
```
areRecordingPermissionsGranted

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
+    {
+        super.onActivityResult(requestCode, resultCode, data);
+        if (RecorderService.REQUEST_CODE_SCREEN_CAPTURE == requestCode) {
+            if (resultCode == Activity.RESULT_OK) {
```
we could use `&&` instead of nested `if`s

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,367 @@
+            MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();
+            long lastAudioTimestampUs = -1;
+
+            while (!stopped && !asyncError) {
```
is it possible to extract audio and video recording to separate private methods to make this one easier to read?

sirmordred replied:
done, https://github.com/sirmordred/io.appium.settings/commit/396e861b2d723258f833c87222448d1c52a10bc6

mykola-mokhnach commented:
AFAIK app sound would only be recorded if the particular app allows that. Also, does the `appops` command below suppress the confirmation dialog that Media projection usually shows before starting the recording?

sirmordred commented:
thank you so much for your review and feedbacks, i'll address them asap i hope, im doing this in my free time and i'll be on vacation for a while, after that i will continue to work on it and try to address your questions etc. thanks

mykola-mokhnach commented:
> thank you so much for your review and feedbacks, i'll address them asap i hope, im doing this in my free time and i'll be on vacation for a while, after that i will continue to work on it and try to address your questions etc. thanks

Sure. We appreciate your contribution

sirmordred commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+            showNotification(TAG); // TODO is this really necessary
+
+            MediaProjectionManager mMediaProjectionManager =
+                    (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
```
done

sirmordred commented:
i am currently updating codebase commit-by-commit but will squash all of them into single commit when all feedbacks are addressed, addressed lots of them now, there are some left that i am currently working on it 👍

sirmordred commented:
> Thank you. Is there any reason why you did not add this feature in https://github.com/appium/appium-uiautomator2-server , for example, as a `mobile:` command? https://github.com/appium/appium-uiautomator2-driver (kind of a basic question as my out of curiosity)
> 
> [updated]
> 
> ah, this was for the below comment in the linked issue?
> 
> > Which means one can simply install any sound recording app, control it using UIA2 to start/stop recording process and be happy about it %)

yeah since recording on android via public android APIs are only available on application level, thats why i targeted io.appium.settings app and for future direction, i think yeah the start/stop recording commands can be folded into much less lenghty commands like mobile: startRecordingEtc.

sirmordred commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,138 @@
+    @SuppressLint("NewApi")
+    @Override
+    public int onStartCommand(final Intent intent, final int flags, final int startId) {
+        Log.v(TAG, "onStartCommand:intent=" + intent);
```
we are checking it on here https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/Settings.java#L88 and if SDK level is lower than 21 this and all further code pieces never gets triggered, the `SuppressLint NewApi `annotations are just for cosmetical purposes but i again include a check if you want 👍

mykola-mokhnach replied:
would it be possible to use `@TargetApi` annotation instead?

sirmordred replied:
done, yeah `@RequiresApi` can be used, i changed all of suppress annotations to `@RequiresApi` now

sirmordred commented:
> AFAIK app sound would only be recorded if the particular app allows that. Also, does the `appops` command below suppress the confirmation dialog that Media projection usually shows before starting the recording?

yeah there are limitations imposed by android for recording, not %100 sure but AFAIK system apps can record audio from apps whether or not app allows it and if we can find a way to trigger recording via adb-only or adb-executed-java-code (with shell process uid/gid 2000) this can mimic system app and can record all audio from all apps (or i am wrong?)

and yeah `appops` command suppress the mediaprojection confirmation dialog 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,143 @@
+
+        int result = START_STICKY;
+        final String action = intent != null ? intent.getAction() : null;
+        Log.v(TAG, "onStartCommand:intent=" + "action string: " + action);
```
extra `+` here?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,143 @@
+
+            if (mMediaProjectionManager != null) {
+                startScreenRecord(mMediaProjectionManager, intent);
+            }
```
should we print an error message instead?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,143 @@
+                final float scaleY;
+                if (width > height) {
+                    // Horizontal
+                    scaleX = width / 1920f;
```
I'm still not quite convinced about these hardcoded values. Not all phones have the same width and height. Perhaps we could read these values from the display manager instead

sirmordred replied:
+1000 yeah that was the solution what i was thinking while updating, i will look into it to make it dynamic 👍

sirmordred replied:
okay done, https://github.com/sirmordred/io.appium.settings/commit/031b55b7a4995dd5a8aaa5c1de3d1089c4b07b13 the problem was that Media Encoder supports different range of width/heights on different phones, so i clamped device specific width/height in supported range to make it dynamic 👍 also i fixed the orientation problem (it was crashing before) it now records in both landscape and portrait modes successfully, but this is somewhat workaround, real fix is rotating video frames before writing it with muxer, since rotating requires matrix transformations etc. it may be slow on android java (i added TODO into code for future) 👍

sirmordred commented:
okay i completed round 2, there are now 2 feedbacks left that i'm currently working too 👍 
https://github.com/appium/io.appium.settings/pull/88#discussion_r867667721
https://github.com/appium/io.appium.settings/pull/88#discussion_r879363641

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderConstant.java:
```diff
@@ -0,0 +1,14 @@
+package io.appium.settings;
+
+public class RecorderConstant {
```
this could be an interface

sirmordred replied:
"constants in class vs constants in interface" is a lot debated topic for a long time AFAIK (https://stackoverflow.com/a/1373019) it seems interface approach is oldschool but i can do it however if that is coding convention in appium 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderConstant.java:
```diff
@@ -0,0 +1,14 @@
+    public static final int AUDIO_CODEC_SAMPLE_RATE = 44100;
+    public static final int AUDIO_CODEC_CHANNEL_COUNT = 1;
+    public static final int AUDIO_CODEC_REPEAT_PREV_FRAME_AFTER = 1000000;
+    public static final int AUDIO_CODEC_I_FRAME_INTERVAL = 5;
```
I usually add unit names to time constants were possible, for example `AUDIO_CODEC_I_FRAME_INTERVAL  -> AUDIO_CODEC_I_FRAME_INTERVAL_MS`

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+package io.appium.settings;
```
Please add the standard header to all newly added files

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+    private static final String TAG = "RecorderService";
+
+    public static final int REQUEST_CODE_SCREEN_CAPTURE = 123;
+    private static final String BASE = "io.appium.settings.recording";
```
can we read the package name dynamically instead of hardcoding it?

sirmordred replied:
it is possible yeah but we have 2 choices AFAIK, one of them is `context.getPackageName()` and other one is `BuildConfig.APPLICATION_ID` and since 1st one requires Context object and we dont have context in that code flow, i can apply 2nd method, is it fine for you ? 👍

sirmordred replied:
okay done https://github.com/sirmordred/io.appium.settings/commit/deff3af08b2254951afa8d28d34c2dcd0197f437

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+    public static final String ACTION_RECORDING_START = BASE + ".ACTION_START";
+    public static final String ACTION_RECORDING_STOP = BASE + ".ACTION_STOP";
+    public static final String ACTION_RECORDING_RESULT_CODE = "result_code";
+    public static final String ACTION_RECORDING_FILENAME = "recording_filename";
```
just `path` would be file

sirmordred replied:
since users will only be allowed to type custom filename of recording file (we dont allow custom paths we allow only custom filename) i thought "filename" is more appropriate but again i can change it if you want  👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    public void onDestroy() {
+        Log.v(TAG, "onDestroy:");
+        if (recorderThread != null && recorderThread.isRecordingContinue()) {
```
isRecordingContinue -> isRecordingRunning or isRecordingInProgress

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+            if (mMediaProjectionManager != null) {
+                startScreenRecord(mMediaProjectionManager, intent);
+            } else {
+                Log.e(TAG, "mMediaProjectionManager null");
```
should we change the result as well?

sirmordred replied:
done, good catch thank you

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+            Log.v(TAG, "onStartCommand:intent=" + "stop");
+            stopScreenRecord();
+            result = START_NOT_STICKY;
+        }
```
should we return an error is the action name is unknown?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void startScreenRecord(MediaProjectionManager mediaProjectionManager,
+                                   final Intent intent) {
+        if (recorderThread == null) {
```
it is much easier to read the code if guards are present there instead of long netsed blocks, e.g.:

```
if (recorderThread == null) {
   ...
} else {
  ....
}
```

could be replaced with 

```
if (recorderThread != null) {
   // print a message and short circuit
   ...
   return;
}
// continue the normal flow
....
```

sirmordred replied:
absolutely right 👍  done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+        if (recorderThread == null) {
+            Log.v(TAG, "recorder thread null");
+
+            final int resultCode = intent.getIntExtra(ACTION_RECORDING_RESULT_CODE, 0);
```
I also like kotlin and scala because they have a dedicated keyword to create a constant variable, but for java it is usually ok to skip `final` modifier unless we really concerned aout the variable value could be unexpectedly changed later

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+            // get MediaProjection
+            final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                    intent);
+            if (projection != null) {
```
same as above

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+                    recorderThread = new RecorderThread(projection, outputFilePath,
+                            rawWidth, rawHeight, isRotated);
+                    recorderThread.startRecording();
+                } else {
```
same as above. Too much of nesting, thus hard to read.

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,135 @@
+                        metrics.widthPixels, metrics.heightPixels, rawWidth, rawHeight));
+                String outputFilePath = intent.getStringExtra(ACTION_RECORDING_FILENAME);
+                if (outputFilePath != null) {
+                    recorderThread = new RecorderThread(projection, outputFilePath,
```
should we also validate if the file name does not contain invalid characters, which cannot be used in linux file names?

sirmordred replied:
AFAIK, there is no "file path validation" method on java (at least on java 8) but i may be wrong, there are other type of validation techniques (like creating empty file for checking) but they seems a bit hacky solutions, i am open to ideas ? 👍 
 thanks

mykola-mokhnach replied:
I tend to use simple regex validation like in https://stackoverflow.com/questions/2679699/what-characters-allowed-in-file-names-on-android

sirmordred replied:
okay done https://github.com/sirmordred/io.appium.settings/commit/dc8dc277448fb3e401559698c10ab17ca8030193 i searched a bit on the net and there were lots of regex that some of them were very restrictive and some does not catch all forbidden chars, so i kanged android's internal validation from aosp :) https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/core/java/android/os/FileUtils.java#995

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+package io.appium.settings;
```
header comment is missing

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+
+    private static final String TAG = "RecorderThread";
+
+    private MediaProjection mediaProjection;
```
consider marking variables that are supposed to only be set once in the constructor as `final`

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+        public void onStopped() {
+            super.onStopped();
+            if (!stopped) {
+                asyncError = true;
```
how about renaming the var to `hasAsyncError` or `didErrorHappenAsync`

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+        this.isRotated = isRotated;
+        videoMime = MediaFormat.MIMETYPE_VIDEO_AVC;
+        bufferInfo = new MediaCodec.BufferInfo();
+        lastAudioTimestampUs = -1;
```
should we instead put the default value to where the var is defined?

mykola-mokhnach replied:
same above

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+    private static final String TAG = "RecorderThread";
+
+    private MediaProjection mediaProjection;
+    private VirtualDisplay virtualDisplay;
```
do we really need to share so many variables across the instance?

sirmordred replied:
i may try to reduce it via making them local vars but there is not so much of that type convertible-vars in it currently if i am not wrong

mykola-mokhnach replied:
I personally find it much more convenient to work with something that looks like

```
public X someMethod() { return new X;}

public void someOtherMethod () {
  X x = someMethod();
}
```

rather than 

```

private x;

private void someMethod() {
   this.x = ...;
}

public void someOtherMethod() {
  someMethod();
  // do something this this.x
}
```

Basically the second one is hard to parse because it assumes an implicit dependency of someOtherMethod on someMethod. If someMethod is not called before someOtherMethod then we'd get an unexpected NPE because of uninitialised `x`. 

The first option makes it more obvious that we need to call someMethod in order to have the instance of X. This becomes critical if there are several variables similar to `x`. The rule of thumb here is to limit the scope of a variable to the necessary minimum and don't expand it unless it is required. Almost the same is also applicable to variables privacy level and to sizes of local variable visibility namespaces.

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void setupVirtualDisplay() {
+        virtualDisplay = mediaProjection.createVirtualDisplay("Appium Screen Recorder",
```
consider this method to return VirtualDisplay instance rather that being void. Having so many depending methods creates a dependency clear where it is not were clear in which order they must be called and when.

sirmordred replied:
done

sirmordred replied:
other `setupXX` methods also starts instance inside them (e.g `setupAudioEncoder` -> `audioEncoder.start()`) should i change them too? (i will put `.start()` calls to outside of function)

mykola-mokhnach replied:
up to you. I've wrote an extended comment above regarding the common variable scoping strategy

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+        int minBufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig,
+                AudioFormat.ENCODING_PCM_16BIT);
+
+        AudioFormat.Builder audioFormatBuilder = new AudioFormat.Builder();
```
I assume this intermediate variable is redundant. We could build a new audio format using the following syntax:

```
AudioFormat newAudioFormat = new AudioFormat.Builder()
  .setEncoding()
  ....
```

mykola-mokhnach replied:
same be

sirmordred replied:
done good catch thank you 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+                        }
+                        inputBuffer.clear();
+                        int read = audioRecord.read(inputBuffer, inputBuffer.capacity());
+                        if (read < 0) {
```
perhaps <= 0 ?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+        audioRecordThread.start();
+    }
+
+    private long getPresentationTimeUs() {
```
this could be simplified by making startTimestampInitialized of type Optional or at least Long. So we can clearly see when it is not initialized yet (e.g. equals Optional.empty() or null)

sirmordred replied:
I couldn't fully understand this one, can you give a detail a bit? thanks 👍

mykola-mokhnach replied:
currently `-1` is used as so-called "guard value". This might be not very obvious as there is no clear indication about why it was selected being magic number.  
To make this more obvious one would make use of Optional. Even making the guard value to a constant would already significantly improve the readability. For example, compare the following:

```
long x = -1;
long x = NO_TIMESTAMP_SET;
Optional<Long> x = Optional.empty(); // then y = x.getOrElseThrow(TimestampNotInitializedError::new)
```

sirmordred replied:
hmm got it but `startTimestampInitialized` is currently boolean value, i couldn't see the  `-1` value anywhere in this `getPresentationTimeUs` method, i now changed name to `isStartTimestampInitialized` and initialized to false as default, renaming helps a bit for readability but i couldn't understand the win by making current `boolean` type to `Optional<boolean>` (we already init as false so no null check or exception-throwing check is required) sorry for misunderstanding

sirmordred replied:
okay done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+        encoderStatus = videoEncoder.dequeueOutputBuffer(bufferInfo,
+                RecorderConstant.MEDIA_QUEUE_BUFFERING_DEFAULT_TIMEOUT);
+        if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
+            if (videoTrackIndex > 0) {
```
this comparison does not match to the log message below

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+
+        encoderStatus = audioEncoder.dequeueOutputBuffer(this.bufferInfo, 0);
+        if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
+            if (audioTrackIndex > 0) {
```
this comparison does not match to the log message below

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+            // set output file orientation info if it is in landscape mode (default is 0==portrait)
+            // note: this method must be run before muxer.start()
+            if (isRotated) {
+                muxer.setOrientationHint(90);
```
cannot we just read the number of rotation degrees from the system? I know there are some devices where 90 corresponds to portrait rather than landscape

sirmordred replied:
done 👍 i changed to dynamic approach, it cost us calling another system service https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/Settings.java#L132 but dynamic over hardcoded is the real solution yeah 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,419 @@
+                    audioRecordThread = null;
+                } catch (InterruptedException e) {
+                    Log.w(TAG, "Error releasing resources, audioRecordThread: ", e);
+                    e.printStackTrace();
```
should we return here?

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +82,98 @@ public void onCreate(Bundle savedInstanceState) {
+    }
+
+    private void handleRecording(Intent openingIntent) {
+        if (openingIntent != null && openingIntent.getAction() != null) {
```
can we use guard conditions to make this method more readable?

sirmordred replied:
done

sirmordred commented:
okay done another round 👍

sirmordred commented:
okay i rewrote most of the parts of the code, refactored according to convention you stated that i also fully agree 👍  https://github.com/appium/io.appium.settings/pull/88#discussion_r889584371

sirmordred commented:
I eliminated lots of class variables and also `this.xx` usage 👍

sirmordred commented:
okay ready 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderService.java:
```diff
@@ -0,0 +1,177 @@
+        final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                intent);
+        if (projection == null) {
+            Log.e(TAG, "MediaProjection is null");
```
such log messages are not very helpful. I usually try to keep them more informative, for example:
Cannot start screen recording: unable to retrieve MediaProjection instance

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/cd56a2634a23e37076ce2f5d5ac2c005d1f52e0e

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderConstant.java:
```diff
@@ -0,0 +1,38 @@
+    public static final String ACTION_RECORDING_STOP = PACKAGE_NAME + ".recording.ACTION_STOP";
+    public static final String ACTION_RECORDING_RESULT_CODE = "result_code";
+    public static final String ACTION_RECORDING_FILENAME = "recording_filename";
+    public static final String ACTION_RECORDING_ROTATION = "recording_rotation";
```
It still makes sense to make values like:
- bitrate
- scale
- FPS 
configurable. They affect the size of the resulting video and we don't always want it to be too huge.

I also usually add an automatic timer to such recorders to prevent the recording to continue forever and eventually exceed the free sdcard space in case of a client failure.

mykola-mokhnach replied:
maybe we also need a separate flag on whether to disable the audio track or video track

sirmordred replied:
> It still makes sense to make values like:
> 
> * bitrate
> * scale
> * FPS
>   configurable. They affect the size of the resulting video and we don't always want it to be too huge.
> 
> I also usually add an automatic timer to such recorders to prevent the recording to continue forever and eventually exceed the free sdcard space in case of a client failure.

well we are not and dont have scaling right now (we are using device's default width/height) scaling will bring complexity since we need to somehow transform images on-the-buffer (it will also slow down recording and may cause jitters probably) this can be future-direction, we can add as a feature on top of recording in the future maybe 👍 
_for bitrate_: since we also calculate bitrate dynamically and use width/height to calculate it, there is no large room for fine-tuning bitrate (only bitrate multiplier which can be dynamic) or else we will allow full customization without boundaries? 
_for fps_: yeah i will take a look at it to make it dynamic let's see how it goes 👍

sirmordred replied:
> I also usually add an automatic timer to such recorders to prevent the recording to continue forever and eventually exceed the free sdcard space in case of a client failure.

+1 good point, instead of timer, what about storage size limitation? when i looked into limitations of storage on Android, i have found out that some says 3GB is the limit https://stackoverflow.com/a/67017220 and another solution is to check avail size dynamically https://stackoverflow.com/a/20329027

mykola-mokhnach replied:
Regarding the features I'm more than ok to do in in different PRs as this one is anyway quite big and it is not a very easy task to go over the whole source code again and again for each review.

sirmordred replied:
> maybe we also need a separate flag on whether to disable the audio track or video track

+1000 yeah i was thinking the same, since we are already writing audio and video buffers separately, we can control it via flags to produce 

- audio-only -> m4a or aac
- video-only -> mp4
- audio-video mixed -> mp4, default

sirmordred replied:
> Altogether @sirmordred I would like to thank you for the great work and all the efforts invested into this PR. Let me know if/when you decide to look for new job opportunities ;-)

thank you so much for your kind answers and great direction and review 💯  it was pleasure for me to work with you 👍  and really thanks for your job opportunities, i really want to work with you guys and contribute to appium and its infrastructure (lets keep connection through mail, i'll mail it asap 👍 )

sirmordred replied:
done, added maximum allowed duration which is default 15 minute but it can be also configurable by user 👍 
 https://github.com/sirmordred/io.appium.settings/commit/afa8fbcef223013cc6c46a9442721085e4e82d18

mykola-mokhnach commented on README.md:
````diff
@@ -321,6 +321,31 @@ and add them to the media library if their MIME types are supported. If the
+
+Start Recording:
+```bash
+adb shell am start -n "io.appium.settings/io.appium.settings.Settings" -a io.appium.settings.recording.ACTION_START --es "recording_filename" "abc.mp4"
````
Please add descriptions for all supported arguments

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/a49f98efb54ac6cca1e6cc3e089fcb596cb5b2d4

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,437 @@
+    private long startTimestampUs;
+    private long lastAudioTimestampUs = NO_TIMESTAMP_SET;
+
+    private int videoTrackIndex = -1;
```
It is better readable when constants are used instead of magic numbers

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,437 @@
+    private final int videoHeight;
+    private final int recordingRotation;
+
+    private boolean muxerStarted;
```
should we also assign default values here?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.Q)
+    private AudioRecord initAudioRecord(MediaProjection mediaProjection, int sampleRate) {
```
I like the way this code is designed 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,437 @@
+        return new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Thread.currentThread().setPriority(Thread.MAX_PRIORITY);
```
this should also probably be configurable. For some people it is more important to have the app under test performing well rather than the video recording

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/d229c009f8bb8e9e00f4dbe70f99b890e3b1aaf7

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,437 @@
+            lastAudioTimestampUs = NO_TIMESTAMP_SET;
+
+            while (!stopped && !hasAsyncError) {
+                if (!writeAudioBufferToFile(audioEncoder, muxer, bufferInfo)) {
```
this is so much easier to read 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/RecorderThread.java:
```diff
@@ -0,0 +1,437 @@
+                    audioRecordThread = null;
+                } catch (InterruptedException e) {
+                    Log.w(TAG, "Error releasing resources, audioRecordThread: ", e);
+                    e.printStackTrace();
```
isn't including error object to the log message already printing its stacktrace?

sirmordred replied:
done

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +93,189 @@ public void onCreate(Bundle savedInstanceState) {
+        if (display != null) {
+            int deviceRotation = display.getRotation();
+
+            switch(deviceRotation) {
```
do we support rotation change in runtime?

sirmordred replied:
not yet but it doesn't crash it as well if user change rotation of device in the middle of recording 👍

mykola-mokhnach commented:
The CLA bot is still complaining about the unsigned CLA. WE will be unable to merge the PR unless it is signed. Could you please verify your local git email address matches the one used to sign CLA?

sirmordred commented:
> The CLA bot is still complaining about the unsigned CLA. WE will be unable to merge the PR unless it is signed. Could you please verify your local git email address matches the one used to sign CLA?

yeah i currently didn't sign EasyCLA but will do as soon as possible 👍  also do you prefer to squash all commits into one-single commit? i can do that 👍 and last, before merging it, do you have a chance to try it on different phones? i tried it several times after each commit i made, on my dev phone which is Huawei P40 Lite running Android 10, it works great 👍 but dont know the situation on 11 and 12 👍

sirmordred commented:
Okay, i got a chance to try it on Samsung A53 running Android 12, it was crashing due to width/height, somehow even though we clamp our width/height corresponding to encoder's capabilities https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/RecorderThread.java#L346 encoder was crashing if width/height larger than 1080x1920 so i set another clamp that act as maximum resolution we support which is 1080x1920 https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/RecorderThread.java#L351

and another obstacle, really big one, was android's latest Storage Access Framework SAF (https://medium.com/androiddevelopers/android-11-storage-faq-78cefea52b7c) which is enforced on Android 11&12 (is permissive on 10)

the old way which we are supplying file absolute path (simply String) into MediaMuxer https://github.com/sirmordred/io.appium.settings/blob/dc8dc277448fb3e401559698c10ab17ca8030193/app/src/main/java/io/appium/settings/RecorderThread.java#L368 is not working on Android SAF, in android SAF, it seems Google blocked raw file access via it's path, also there is no problem in application-specific-directories(`/data/data/com.xx` or `/sdcard/Android/data/com.xx`) in SAF but `adb` is not able to pull file from application's-specific-directories on SAF anymore (permission denied), so i needed to find a common path across different phones and android versions, which must be writable-by-application-itself and plus it must be pullable-by-adb, so i came up with `MediaStore.Downloads.EXTERNAL_CONTENT_URI` https://github.com/sirmordred/io.appium.settings/commit/9177d64d487d94e3773552652174587675425b09#diff-0d9c9092f6e43c01e7c8a51982fb0c221bcb9f3944fdf0349d792453df871a22R199  (there is also no file path terminology on SAF mostly `Uri`s) the real path of that Uri is `phone's external storage(sdcard) + Downloads folder + OurVideoFilename.mp4` or `/storage/emulated/0/Downloads/AppiumRecord.mp4` that file path is both writable-by-app and pullable-by-adb, and with these updates, recording started to work great on Android 12 too 👍 as far as i tested it on Samsung A53

but there is a problem, since absolute path of `MediaStore.Downloads.EXTERNAL_CONTENT_URI` Uri which we are saving output videos into it, do not correspond real path or MAY CHANGE across different android phones and versions (i am not %100 sure about that but it is possible i think), we need to find a path dynamically inside app itself and give it to user instead of hardcoded path like `/storage/emulated/0/Downloads/AppiumRecord.mp4` 👍 and we can probably do it without problem via `adb` and `broadcast intents` like we already do for Location data https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/receivers/LocationInfoReceiver.java#L38 but at cost, we will require another command to all users of this feature :/ (START, STOP, GETPATH, PULL)
also we should keep in mind that getting real path from Uri is somehow not so official way :/ take a look https://gist.github.com/tatocaster/32aad15f6e0c50311626 (but i tried it on Samsung Android 12 and it worked good)

sorry for that long explanation but unfortunately that is because Google really messed up storage on latest version of android :)

What do you think, thanks

mykola-mokhnach commented:
I think you could still pull files from app directories using adb using one small trick: https://github.com/appium/appium-android-driver/blob/89da9a1728c39aeceae18a1eb8f04e7bf0ba9602/lib/commands/file-actions.js#L93
The key here is to call run-as

sirmordred commented:
Hmm i wasn't aware of that thanks 👍  also while investigating more different storage paths, i realized that old way is still working on android 12, i mean https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/Settings.java#L141 but in the first run, it was not working (or i am wrong?), after some investigation of that method, i learned that in the 1st call of that method `getExternalFilesDir(null)` OS creates a subfolder in the external storage for us `/storage/emulated/0/Android/data/io.appium.settings/files` but it seems subfolder creation does not work on 1st call of that method sometimes, so it needs to be called twice (as per https://www.androidbugfix.com/2021/10/getexternalfilesdirnull-returns-null-in.html) so i double called it and it now works perfectly 👍 

other changes:

- I disabled the "default file name" logic in case of user typed invalid filename because it creates confusion for user i mean user typed `a"/.mp4` but instead of exiting, we still continue and create record file with `Appium_2020xx.mp4`  after record, user will not find recorded file in an expected path `/storage/xx/a"/.mp4`  but instead `/storage/xx/Appium_2020xx.mp4` https://github.com/sirmordred/io.appium.settings/commit/f6624c58686e351e83bba6a7aa2294993ccdbbfc
- I moved some of the helper methods into its own util class to not pollute Settings class https://github.com/sirmordred/io.appium.settings/commit/f878b4952ac850549b1414d7a311650ef4c67de6
- After creating also RecorderUtil class, we now have 4 class for recording feature, so to organize it more, i moved them into separate sub-folder https://github.com/sirmordred/io.appium.settings/commit/deae0f79e0357d87fccca216073f7db9d235b2e4 they deserve it :)
- Addressed some of your last feedbacks 👍  https://github.com/sirmordred/io.appium.settings/commit/d1ae86a57fff381f4fd5469fcc818001cb16e3bb

now there are 2 feedbacks left:
- https://github.com/appium/io.appium.settings/pull/88#discussion_r889719451
- https://github.com/appium/io.appium.settings/pull/88#discussion_r889719965

after addressing them, we can revisit the --es parameters (there is 1 param now which is --es recording_filename) which we will expose to users and to whole internet :) before adding into official appium app, i am %100 open to any name/parameter changes, replacements 👍

sirmordred commented:
Also good news is we probably can eliminate the need of `adb shell pm grant WRITE_EXTERNAL_STORAGE` because we are already writing data to app-specific-directory (`/storage/emulated/0/Android/data/io.appium.settings/files`) 👍 
Refs:

- https://stackoverflow.com/a/34469525
- https://developer.android.com/training/data-storage

sirmordred commented:
Okay fixed all feedbacks 👍

sirmordred commented:
there can be another round of feedback no problem 👍 (We may revisit argument names i.e `recording_filename` `recording_max_duration` etc they seems a bit lengthy so i can re-name these to anything you want 👍 ) and if everything is okay, i can squash all commits into one single commit and then  -> sign EasyCLA to prepare for merge 👍

mykola-mokhnach commented on README.md:
````diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+
+```bash
+adb shell pm grant io.appium.settings android.permission.RECORD_AUDIO
+adb shell pm grant io.appium.settings android.permission.WRITE_EXTERNAL_STORAGE
````
is this permission still needed?

sirmordred replied:
done, removed it https://github.com/sirmordred/io.appium.settings/commit/f0b2628ec6eed40ddf1e880a730747f987d39331, its not needed anymore 💯 as far as i  tested it on both Huawei Android 10 and Samsung Android 12

mykola-mokhnach commented on README.md:
````diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+
+Start Recording:
+```bash
+adb shell am start -n "io.appium.settings/io.appium.settings.Settings" -a io.appium.settings.recording.ACTION_START --es "recording_filename" "abc.mp4" --es "recording_priority" 0 --es "recording_max_duration" 900
````
I would rather go with simple arg names, without the `recording_` prefix. Also `max_duration_sec` would be more precise

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/2c8878a8387a0b425117abfd6db12a12292b25a0

mykola-mokhnach commented on README.md:
```diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+
+### Arguments (see above start command as an example for giving arguments)
+- recording_filename (Mandatory) - You can type recording video file name as you want, but recording currently supports only "mp4" format so your filename must end with ".mp4"
+- recording_priority (Optional) - Default value: 0 which means recording thread priority is Maximum hovewer if you face performance drops during testing with recording enabled, you can reduce recording priority to 2 - Normal or 1 - Minimum
```
shouldn't 1 be normal and 2 minimum if 0 is the high priority?

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/2c8878a8387a0b425117abfd6db12a12292b25a0 yeah also to be consistent with the docs (...def is max, you can reduce priority...) i changed max(default) -> 2, normal -> 1 and min -> 0

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +94,146 @@ public void onCreate(Bundle savedInstanceState) {
+        }
+
+        if (recordingAction.equals(ACTION_RECORDING_START)) {
+            String recordingFilename = intent.getStringExtra(ACTION_RECORDING_FILENAME);
```
can it be null?

sirmordred replied:
yeah it can be but we are checking against it on here https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/recorder/RecorderUtil.java#L57 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +94,146 @@ public void onCreate(Bundle savedInstanceState) {
+                finishActivity();
+                return;
+            }
+            recordingOutputPath = externalStorageFile.getAbsolutePath()
```
you could also use https://stackoverflow.com/questions/711993/does-java-have-a-path-joining-method or https://developer.android.com/reference/java/nio/file/Paths

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/991c57ebd035ea691191c3cf60a01d414cd867ea

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +94,146 @@ public void onCreate(Bundle savedInstanceState) {
+            return;
+        }
+
+        if (recordingAction.equals(ACTION_RECORDING_START)) {
```
what do we do if another recording is currently running?

sirmordred replied:
we simply inform user via log and exiting https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/recorder/RecorderService.java#L115 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/recorder/RecorderConstant.java:
```diff
@@ -0,0 +1,48 @@
+    public static final int AUDIO_CODEC_DEFAULT_BITRATE = 64000;
+    public static final int VIDEO_CODEC_FRAME_RATE = 30;
+    public static final long MEDIA_QUEUE_BUFFERING_DEFAULT_TIMEOUT_MS = 10000;
+    public static final long NANOSECOND_TO_MICROSECOND = 1000;
```
NANOSECONDS_IN_MICROSECOND

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/f618ac9cf905ccc87157d24f2e685566bd703612

mykola-mokhnach commented on app/src/main/java/io/appium/settings/recorder/RecorderService.java:
```diff
@@ -0,0 +1,192 @@
+
+        int recordingRotation = intent.getIntExtra(ACTION_RECORDING_ROTATION, -1);
+
+        if (recordingRotation == -1) {
```
can we get the orientation value from resources: https://github.com/appium/appium-uiautomator2-server/blob/5d49661b6198a52c59e869005a66e9a0f589c432/app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java#L51 ?

sirmordred replied:
we can probably use but in my opinion our current method handles orientation more accurate and with more coverage (with exact 4 modes, 0, 90, 180 and 270) ref: https://stackoverflow.com/a/3663693

mykola-mokhnach commented on app/src/main/java/io/appium/settings/recorder/RecorderThread.java:
```diff
@@ -0,0 +1,466 @@
+                                                int videoHeight, int videoBitrate) {
+        MediaFormat encoderFormat = MediaFormat.createVideoFormat(videoMime, videoWidth,
+                videoHeight);
+        encoderFormat.setInteger(MediaFormat.KEY_COLOR_FORMAT,
```
If setInteger returns `this` then we could simply chain all calls:

```
return MediaFormat.createVideoFormat(videoMime, videoWidth, videoHeight)
  .setInteger()
  ....
```

sirmordred replied:
i would have love to do this but unfortunately `setInteger` is void type function

sirmordred commented on README.md:
````diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+
+Obtain Recording Output File:
+```bash
+adb pull /storage/emulated/0/Android/data/io.appium.settings/files/abc.mp4 abc.mp4
````
yeah as far as i tested it on both phones with different android versions (huawei android 10 and samsung android 12) the path is both writable-by-app and pullable-by-adb without problem 👍

sirmordred commented on app/src/main/java/io/appium/settings/recorder/RecorderConstant.java:
```diff
@@ -0,0 +1,48 @@
+    public static final int RECORDING_PRIORITY_MIN = 1;
+    public static final int RECORDING_PRIORITY_NORM = 2;
+    public static final int RECORDING_PRIORITY_DEFAULT = Thread.MAX_PRIORITY;
+    public static final int RECORDING_MAX_DURATION_DEFAULT_MS = 15 * 60 * 1000; // 15 Minute, in miliseconds
```
done https://github.com/sirmordred/io.appium.settings/commit/2c8878a8387a0b425117abfd6db12a12292b25a0

sirmordred commented:
okay while i was testing recording in different emulators (even if it worked perfect on real devices, Huawei Android 10 and Samsung Android 12) it crashed on emulators (media encoder was crashing while configuring it with given width/height resolutions) and i realized that this is because media codecs and its capabilities as well as device display resolutions vary device-to-device very much and the core problem is media encoder/decoder (encoder in our case) is not GUARANTEED to encode/decode in given any resolutions (width/height) 

in the past, we were giving raw display width/height to media encoder https://github.com/sirmordred/io.appium.settings/blob/203ef2dd68231f69d368a3a64a1909f614f3aaa8/app/src/main/java/io/appium/settings/recorder/RecorderThread.java#L372 of course we were clamping it, but clamp was not not working reliably, even if we clamp width/height against media encoder's capabilities, it was resulting values like 1080x1823 (some devices have notch or virtual back/home button fields some not) and media encoder was crashing with these type of values, 

so i deep dive into media encoder and its capabilities, i saw that media encoder can support up to 5 different resolution modes, i got that assumption from CTS tests, see https://github.com/aosp/cts/blob/08e720f8ee9a1170d995b03b748528098f22e101/tests/tests/media/src/android/media/cts/VideoEncoderTest.java#L1346 and above,  since all released android phones MUST pass CTS tests, we can safely assume that media encoder can support these resolution modes, so i added these resolution modes and helper functions to detect supported max resolution on-device, and media encoder is now works without crashing as far as i tested it on both real-devices and emulators 👍  i also added user-selectable-resolution-mode argument to making it dynamic (didnt test different resolutions yet, only tested max supported resolution on device, but will test it ASAP 👍 )

other changes:
- i was giving display density as hardcoded while configuring `VirtualDisplay`  https://github.com/sirmordred/io.appium.settings/commit/4a0e029fdf26e942b38926894686be3333ee24da#diff-423e8dd14708c0aa3b8df6ff5c75cd8a780ea48625ac50a32a194b167cfe1259R138 so i changed it to dynamic since we can get value from device directly
- I added additional control that checks if media projection permission is granted or not (user typed `adb shell appops project_media allow` command or not) https://github.com/sirmordred/io.appium.settings/commit/203ef2dd68231f69d368a3a64a1909f614f3aaa8 to make recording more solid (tested it on different phones and works great) 👍

sirmordred commented:
Voilaa 💯  i tested different resolution modes and it works perfect 👍 and resulting output videos are really smaller in size (as expected) we now support different resolutions 👍

mykola-mokhnach commented:
AFAIK the AVC video encoder used here supports pretty wide set of resolutions. One of main requirements to these resolutions is, though, that both width and height must be [divisible](https://stackoverflow.com/questions/20847674/ffmpeg-libx264-height-not-divisible-by-2) by two. You could also use https://stackoverflow.com/questions/44902334/how-to-get-max-resolution-of-mediacodec-before-configuring to check if the particular width, height and frame rate are supported by the selected codec instead of hardcoding the limits. Usually one  would only provide the scale value to keep the original aspect ratio rather than both width and height

mykola-mokhnach commented:
In iOS, for example, we set the scale value to 2, which means the video resolution is by default half of the original device resolution

sirmordred commented:
> AFAIK the AVC video encoder used here supports pretty wide set of resolutions. One of main requirements to these resolutions is, though, that both width and height must be [divisible](https://stackoverflow.com/questions/20847674/ffmpeg-libx264-height-not-divisible-by-2) by two. You could also use https://stackoverflow.com/questions/44902334/how-to-get-max-resolution-of-mediacodec-before-configuring to check if the particular width, height and frame rate are supported by the selected codec instead of hardcoding the limits. Usually one would only provide the scale value to keep the original aspect ratio rather than both width and height

i thought the same but in android's AVC/H264 codec, i saw and experienced supported resolutions are not really that much wide, for example try the following code on any high-end phone or low-end phone (it doesn't matter)

`try {
    String videoMime = MediaFormat.MIMETYPE_VIDEO_AVC;

    MediaCodec videoEncoder = MediaCodec.createEncoderByType(videoMime);

    MediaCodecInfo.VideoCapabilities videoEncoderCapabilities = videoEncoder.getCodecInfo().
            getCapabilitiesForType(videoMime).getVideoCapabilities();

    boolean a1 = videoEncoderCapabilities.isSizeSupported(176, 144);
    boolean a2 = videoEncoderCapabilities.isSizeSupported(320, 240);
    boolean a3 = videoEncoderCapabilities.isSizeSupported(720, 480);
    boolean a4 = videoEncoderCapabilities.isSizeSupported(1280, 720);
    boolean a5 = videoEncoderCapabilities.isSizeSupported(1920, 1080);
    boolean a6 = videoEncoderCapabilities.isSizeSupported(2560, 1440);


    Log.e("OGIX", a1 ? "yes1" : "no");
    Log.e("OGIX", a2 ? "yes2" : "no");
    Log.e("OGIX", a3 ? "yes3" : "no");
    Log.e("OGIX", a4 ? "yes4" : "no");
    Log.e("OGIX", a5 ? "yes5" : "no");
    Log.e("OGIX", a6 ? "yes6" : "no");
} catch (Exception e) {
    e.printStackTrace();
}`

it will print "yes" for all resolutions from 176x144 to 1920x1080 range but it will print "no" for the last one -native 2K resolution 2560x1440- even though it is divisible by 2, so we are much likely limited to these resolutions i guess

sirmordred commented:
Okay i have done everything in my mind/which i noted 👍 also tried on the following devices and i can confirm its working good without problem 👍 

- Huawei P40 Lite running Android 10
- Samsung A53 running Android 12
- Xiaomi MI5 running Android 12 (customized ROM, LineageOS)
- [Emulator] Pixel C Tablet running Android 11
- [Emulator] Pixel 2 Phone running Android 11
Nearly all devices have also different resolutions and dpi as well as display sizes

mykola-mokhnach commented on README.md:
```diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+
+### Arguments (see above start command as an example for giving arguments)
+- filename (Mandatory) - You can type recording video file name as you want, but recording currently supports only "mp4" format so your filename must end with ".mp4"
+- priority (Optional) - Default value: 2 which means recording thread priority is Maximum however if you face performance drops during testing with recording enabled, you can reduce recording priority to 1 - Normal or 0 - Minimum
```
maybe it even makes more sence to provide priority as a string (e.g. high/normal/low)

sirmordred replied:
good idea 👍 done, i change now to string (high/normal/low) https://github.com/sirmordred/io.appium.settings/commit/a61a988c1d59a75e98680a12ca49372d77c73c52

mykola-mokhnach commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +99,152 @@ public void onCreate(Bundle savedInstanceState) {
+            return;
+        }
+
+        if (recordingAction.equals(ACTION_RECORDING_START)) {
```
would `switch` be more readable here?

sirmordred replied:
there are only 3 case (ACTION_START, ACTION_STOP and Unknown action) so `if-elif `or `switch-case `wouldn't make so much difference i guess, and also with `switch-case` we need to indent code blocks extra 1-tab right compared to `if-elif`

```
if (recordingAction.equals(ACTION_START)) {
    blablabla..
}
```

versus 

```
switch (recordingAction) {
    case ACTION_START:
        blablabla
        break;
}
```

but again i can change it if you want/prefer 👍

mykola-mokhnach commented on README.md:
```diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+- priority (Optional) - Default value: 2 which means recording thread priority is Maximum however if you face performance drops during testing with recording enabled, you can reduce recording priority to 1 - Normal or 0 - Minimum
+- max_duration_sec (Optional) (in seconds) - Default value: 900 seconds which means maximum allowed duration is 15 minute, you can increase it if your test takes longer than that
+- resolution_mode (Optional) - Default value: 5 which means default resolution for recording is Full HD 1920x1080 however you can change it to following resolutions: 4 - HD (1280x720), 3 - 480P (720x480), 2 - QVGA (320x240), 1 - QCIF (176x144)
+
```
Ideas for the future:
It would be nice to have FPS value configurable

sirmordred replied:
+1 yeah, noted, will try to sum up and write as comment all future-directions after the PR 👍

mykola-mokhnach commented on app/src/main/java/io/appium/settings/recorder/RecorderConstant.java:
```diff
@@ -0,0 +1,77 @@
+    public static final int RECORDING_PRIORITY_DEFAULT = Thread.MAX_PRIORITY;
+    public static final int RECORDING_MAX_DURATION_DEFAULT_MS = 15 * 60 * 1000; // 15 Minutes, in milliseconds
+    // Pair.create(int width, int height)
+    public static final List<Pair<Integer, Integer>> RECORDING_RESOLUTION_LIST =
```
it would be nice to have an explanation comment here about why we are limiting the list of possible reslutions

sirmordred replied:
done https://github.com/sirmordred/io.appium.settings/commit/cf3aa4755df3e95d1353e8cd8a3e69b88bc37c80

mykola-mokhnach commented on app/src/main/java/io/appium/settings/recorder/RecorderConstant.java:
```diff
@@ -0,0 +1,77 @@
+    // Pair.create(int width, int height)
+    public static final List<Pair<Integer, Integer>> RECORDING_RESOLUTION_LIST =
+            Arrays.asList(
+                    Pair.create(1920, 1080),
```
Pair might not be very readable for such purpose. I would rather have a separate data object for resolution values, for example https://developer.android.com/reference/android/util/Size

sirmordred replied:
great catchup 💯 thanks, i forgot the `Size` yeah its a perfect fit for this use-case 👍  done https://github.com/sirmordred/io.appium.settings/commit/1326e79ca4caf5e34dd3fa198c9b11272f7ce135

mykola-mokhnach reviewed (approved):
had some minor comments though

sirmordred commented on app/src/main/java/io/appium/settings/Settings.java:
```diff
@@ -70,12 +99,152 @@ public void onCreate(Bundle savedInstanceState) {
+                    .toAbsolutePath()
+                    .toString();
+
+            recordingRotation = RecorderUtil.getDeviceRotationInDegree(getApplicationContext());
```
you mean values like 1920x1920 or 4096x4096 right? yeah that is good idea and can potentially fit both orientations as you said 👍  noted, will try to sum up and write as comment all future-directions after the PR 👍

sirmordred commented on README.md:
```diff
@@ -321,6 +321,36 @@ and add them to the media library if their MIME types are supported. If the
+- filename (Mandatory) - You can type recording video file name as you want, but recording currently supports only "mp4" format so your filename must end with ".mp4"
+- priority (Optional) - Default value: 2 which means recording thread priority is Maximum however if you face performance drops during testing with recording enabled, you can reduce recording priority to 1 - Normal or 0 - Minimum
+- max_duration_sec (Optional) (in seconds) - Default value: 900 seconds which means maximum allowed duration is 15 minute, you can increase it if your test takes longer than that
+- resolution_mode (Optional) - Default value: 5 which means default resolution for recording is Full HD 1920x1080 however you can change it to following resolutions: 4 - HD (1280x720), 3 - 480P (720x480), 2 - QVGA (320x240), 1 - QCIF (176x144)
```
you mean we should enforce resolution arg as Mandatory ? or passing resolution arg as string (full_hd, hd, 480p etc.) instead of integer?

mykola-mokhnach replied:
pass a string instead of integer, e.g. 640x480 or 1920x1080

sirmordred replied:
okay done https://github.com/sirmordred/io.appium.settings/commit/bc9ba910056d7b790312a501ddf0bd738b4ea1ea reason i couldn't use `switch-case` in here https://github.com/sirmordred/io.appium.settings/commit/bc9ba910056d7b790312a501ddf0bd738b4ea1ea#diff-f3d89bcfbd433c431567a1fc5306d1e60589ae7abfc5add09b4e0bbcb87cbe66R256 is that it doesn't support android.util.Size object 👍 also shortened argument name `resolution_mode` to `resolution` 👍

sirmordred commented:
thank you so much for the review @mykola-mokhnach 👍 

after resolving last feedback https://github.com/appium/io.appium.settings/pull/88#discussion_r897789127 and receive LGTM from you 👍 i'll squash all commits into single one (not doing this for now because of information loss of commits) and sign EasyCLA 👍

sirmordred commented:
Okay i squashed all changes into single commit but individual commit history can be seen through here (i opened another branch) https://github.com/sirmordred/io.appium.settings/tree/recording_feature and also signed EasyCLA and after all changes, i double-checked that it runs without problem on my real test phones 👍

sirmordred commented:
FUTURE DIRECTIONS:

- Add user selectable argument/flag to support the following modes separately A) **Audio-only recording** B) **Video-only recording** C) **Audio/Video Mixed recording(currently supported)**
- Add user selectable FPS argument/flag to dynamically control frame rate
- Add real orientation support with runtime rotation change handled(by manipulating/rotating frames via GL/MatrixTransformer utils) (Side note: setup the maximum video resolution as (max(W, H) x max(W, H)), which could potentially fit both portrait and landscape orientation modes)
- Add and implement START/STOP/PULL commands into appium-android-driver https://github.com/appium/appium-android-driver to fold and shorten current start/stop commands and make life easier for appium users :)
- Change activity-initiated-recording `adb shell am start xxx` into broadcast-initiated-recording `adb shell am broadcast xxx` to eliminate current activity open/close requirement (this is currently impossible AFAIK because result intent data of `MediaProjectionManager.createScreenCaptureIntent()` must be retrieved through `onActivityResult` and only-activity-classes can define, initiate and handle `onActivityResult`)
- Try to support larger resolutions than Full HD 1920x1080 (current maximum supported resolution)
